include/votca/csg/interaction.h

```diff
@@ -25,71 +25,89 @@
 
 // Local VOTCA includes
 #include "bead.h"
-#include "topology.h"
 
 namespace votca {
 namespace csg {
 
+class Topology;
 /**
     \brief base class for all interactions
 
-    This is the base class for all interactions.
+    This is the base class for all interactions. Interactions class is
+    responsible for describing the bonds between atoms or coarse grained beads
+    it is specifically designed to so that only the topology class can
+    create interactions. The topology class is responsible for the memory
+    management of all the interactions.
 
     \todo double names/groups right, add molecules!!
 */
 class Interaction {
  public:
-  Interaction() = default;
-
   virtual ~Interaction() = default;
-  virtual double EvaluateVar(const Topology &top) = 0;
+  virtual double EvaluateVar(const Topology &top) const = 0;
 
-  std::string getName() const { return _name; }
+  std::string getName() const noexcept { return _name; }
 
-  void setGroup(const std::string &group) {
-    _group = group;
-    RebuildName();
-  }
-  const std::string &getGroup() const {
+  const std::string &getGroup() const noexcept {
     assert(_group.compare("") != 0);
     return _group;
   }
 
   // the group id is set by topology, when interaction is added to it
   // \todo if the group name is changed later, group id should be updated by
   // topology
-  Index getGroupId() {
+  Index getGroupId() const noexcept {
     assert(_group_id != -1);
     return _group_id;
   }
-  void setGroupId(Index id) { _group_id = id; }
 
-  void setIndex(const Index &index) {
-    _index = index;
-    RebuildName();
-  }
-  const Index &getIndex() const {
+  const Index &getIndex() const noexcept {
     assert(_index != -1);
     return _index;
   }
 
-  void setMolecule(const Index &mol) {
-    _mol = mol;
-    RebuildName();
-  }
-  const Index &getMolecule() const {
+  const Index &getMolecule() const noexcept {
     assert(_mol != -1);
     return _mol;
   }
 
-  virtual Eigen::Vector3d Grad(const Topology &top, Index bead) = 0;
+  virtual Eigen::Vector3d Grad(const Topology &top, Index bead) const = 0;
   Index BeadCount() const { return _beads.size(); }
   Index getBeadId(Index bead) const {
     assert(bead > -1 && boost::lexical_cast<size_t>(bead) < _beads.size());
     return _beads[bead];
   }
 
  protected:
+  void setIndex(const Index &index) {
```

Comment:
What is the intention of hiding these methods?

Reply:
It reduces scope, better encapsulation.

Reply:
I wish there was a way to only expose some methods to certain classes.